### ISSUE TITLE
Redirect public link to public Webdav URL

### ISSUE THREAD

**Issue by PVince81** [owncloud/core#21787]
To be able to have resumable URLs, users should use the public Webdav URL which is:
http://$token@host/owncloud/public.php/path/to/file.dat

However the UI is still displaying the old download link which looks like "http://host/owncloud/s/$token/path/to/file.dat&download" which is not resumable.

I suggest that we keep the nicer looking URL in the web UI, but make it redirect to the actual Webdav URL. This is only for single files, not zip bundles.

@DeepDiver1975 @jancborchardt @rullzer @LukasReschke

**Comment by rullzer** [comment]
Sounds like a plan!

**Comment by nickvergessen** [comment]
Just so we don't fail, we have smashbox tests for downloading from those links (should be integration tests, but well we didn't have them back then). So should test if they still pass after the patch.

**Comment by LukasReschke** [comment]
Sounds good for me. :+1:

**Comment by jancborchardt** [comment]
:+1:

**Comment by PVince81** [comment]
Great, thanks for the comments, that counts as a green light :smile:

**Comment by PVince81** [comment]
a cool side effect of this is that the end download URL will also have the real file name, which will hopefully fix the issue where people complained that wget/curl didn't have the proper file name.

**Comment by nickvergessen** [comment]
- [ ] Ensure that activities are still triggered correctly

**Comment by PVince81** [comment]
Note from @nickvergessen in the linked ticket: if we do a redirect it needs "double effort" because we'd need to load all apps twice, once in the "index.php/s/$token" and then again for the public.php link.

**Comment by oparoz** [comment]
Just a note that the webdav endpoint doesn't seem to work currently, at least not with images. Maybe fix it first?
Let me know if this not known and needs to be opened separately first.
I get a link like this:
`public.php/webdav/800px-Hamburg_City_Beach_Club.jpg`
I get a 200, but Chrome says:
>This page contains the following errors:
error on line 1 at column 1: Document is empty
error on line 1 at column 1: Encoding error
Below is a rendering of the page up to the first error.

**Comment by PVince81** [comment]
@oparoz not known, can you make a separate ticket ?
That URL is correct but must also include the token as "http://token@host/owncloud/public.php/webdav/pic.jpg". If it doesn't, then it's a bug.

**Comment by rullzer** [comment]
Still I feel it should not return 200 :P

**Comment by oparoz** [comment]
OK, filed here: https://github.com/owncloud/core/issues/22296

**Comment by PVince81** [comment]
Redirecting from the share link + download to the public webdav will not work, see what firefox says:
![firefox-redirect-token](https://cloud.githubusercontent.com/assets/277525/12956664/63520e10-d029-11e5-8ad2-a43d5eaeaf8d.png)

I think I'll revert the download URL from public.js back to the original non-webdav one for now.

With dav2 we should look into making more unique URLs that don't require a http username but have the token inside them. For example: "remote.php/dav/public/$token/path/to/file" and still be unique. CC @DeepDiver1975

**Comment by PVince81** [comment]
This reverts the links back to non-webdav download links for now: https://github.com/owncloud/core/pull/22297

It's sad because the webdav link made it possible to resume downloads / seek.
I'll raise a ticket to look into alternatives.

**Comment by PVince81** [comment]
If anyone is interested to investigate the redirect case, here's the raw change in sharecontroller.php:
```php
diff --git a/apps/files_sharing/lib/controllers/sharecontroller.php b/apps/files_sharing/lib/controllers/sharecontroller.php
index 08679c8..deb2057 100644
--- a/apps/files_sharing/lib/controllers/sharecontroller.php
+++ b/apps/files_sharing/lib/controllers/sharecontroller.php
@@ -398,6 +398,10 @@ class ShareController extends Controller {
                                ->setAffectedUser($share->getShareOwner())
                                ->setObject('files', $share->getNode()->getId(), $userFolder->getRelativePath($share->getNode()->getPath()));
                        $this->activityManager->publish($event);
+
+                       // redirect to public webdav endpoint for a more direct access (streaming, etc)
+                       $publicDownloadUrl = \OC::$WEBROOT . '/public.php/webdav/';
+                       return new RedirectResponse($publicDownloadUrl);
                }
                // Directory share
                else {
@@ -446,6 +450,21 @@ class ShareController extends Controller {
 
                                        $this->activityManager->publish($event);
                                }
+
+                               if (count($files_list) === 1) {
+                                       $urlPath = '';
+                                       $file = $files_list[0];
+                                       $path = trim($path, '/');
+                                       if (!empty($path)) {
+                                               $urlPath = $path . '/';
+                                       }
+                                       $urlPath .= $file;
+
+                                       // redirect to public webdav endpoint for a more direct access (streaming, etc)
+                                       $publicDownloadUrl = $this->request->getServerProtocol() . '://' . $token . '@' . $this->request->getServerHost() .
+                                               \OC::$WEBROOT . '/public.php/webdav/' . $urlPath;
+                                       return new RedirectResponse($publicDownloadUrl);
+                               }
                        } else {
                                // The folder is downloaded
                                $event = $this->activityManager->generateEvent();
```

**Comment by oparoz** [comment]
Regarding the message posted by Firefox, is there a way to customise it? The main issue I have with it is that it could mean anything. A user could see that as some shenanigans triggered by malware per example.

**Comment by PVince81** [comment]
Customize in which way ? I think the message appears because we're redirecting from a URL that has no username in it to one that has it, so like you said it might be to detect shenanigans.

**Comment by PVince81** [comment]
Moving to 9.1 to keep the discussion/research going.

**Comment by oparoz** [comment]
> Customize in which way ?

I was wondering if there was a way to anticipate the browser's move by maybe creating our own dialogue which would then send the correct information to the browser, but dav2 is probably a better option for a future version of ownCloud.

**Comment by PVince81** [comment]
To summarize the problems so far:
- cannot always use public Webdav URL for download because of "token as username" in URL which isn't available in some APIs like video players
- cannot always use public Webdav URL with "Authorization" header because some APIs like video players don't support passing headers either
- "download" activities are not triggered by Webdav URL, also we don't want to because fed shares would cause a lot of download activities to be triggered

However, to make video player and range requests work with the JS video players, we need to either:
- sort out the Webdav URL problem, possibly by changing the format in dav2 to contain all required info (token)
or
- implement Range request support for the old download endpoint...

**Comment by PVince81** [comment]
At this point I'd rather tend to want to implement range request support in the current non-webdav download endpoint until we get some time to properly design and migrate the dav v2 endpoint.

Objections ? @DeepDiver1975 @icewind1991 @LukasReschke @rullzer

**Comment by icewind1991** [comment]
Sounds good

**Comment by rullzer** [comment]
@PVince81 Yes agreed.

**Comment by PVince81** [comment]
In the end this didn't work out. We had to go back to using another endpoint for downloads and range requests.

And download tracking needs to go to a special URL in order to have it registered only once and only when downloaded over the web UI, not over federated shares.
